src/include/storage/data_table.h

```diff
@@ -88,31 +114,39 @@ class DataTable {
    private:
     friend class DataTable;
 
-    /** Indicates that the iterator should advance to the end of the table. */
-    static constexpr int32_t ADVANCE_TO_THE_END = -1;
-    /** An invalid TupleSlot. */
-    static TupleSlot InvalidTupleSlot() { return TupleSlot(nullptr, 0); }
+    SlotIterator() : is_end_(true) {}
 
-    /**
-     * @warning MUST BE CALLED ONLY WHEN CALLER HOLDS LOCK TO THE LIST OF RAW BLOCKS IN THE DATA TABLE
-     */
-    SlotIterator(const DataTable *table, uint32_t block_index, int32_t num_advances, uint32_t offset_in_block)
-        : table_(table), block_index_(block_index), num_advances_(num_advances) {
-      current_slot_ = {block_index >= table_->blocks_.size() ? nullptr : table->blocks_[block_index], offset_in_block};
-      TERRIER_ASSERT((current_slot_.GetBlock() == nullptr && current_slot_.GetOffset() == 0) ||
-                         current_slot_.GetBlock() != nullptr,
-                     "Offset should be 0 when block is nullptr.");
+    SlotIterator(const DataTable *table) : table_(table), block_index_(0), is_end_(false) {  // NOLINT
+      end_index_ = table_->blocks_size_;
+      TERRIER_ASSERT(end_index_ >= 1, "there should always be at least one block");
+
+      // only 1 empty block
+      if (end_index_ == 1) {
+        common::SharedLatch::ScopedSharedLatch latch(&table_->blocks_latch_);
+        if (table_->blocks_[0]->GetInsertHead() == 0) {
+          end_index_ = 0;
+        }
+      }
```

Comment:
Similar to [before](https://github.com/cmu-db/noisepage/pull/850#discussion_r500705055) the table by default initializes 1 block to start, but there are no tuples in that block. So we need to be able to identify that the slot was created on an empty table with 1 block. So to deal with that we do, otherwise its not equal to end. I thought about maybe putting this logic in `==` but that is called very frequently so I figured its best to keep it streamlined?

Reply:
This is special casing only one instance of a much larger case.  I think the right way to do this would be record the insert head for the block when you first look at it and only iterate to that point since any concurrent insertions are logically invisible to this thread.  As long as we check the stored tupleslot validity on initialization and advancement and set to `end` when appropriate, then we should be able to consolidate this logic into a single location.

Reply:
done